pkgs/cupertino_http/lib/src/cupertino_web_socket.dart

```diff
@@ -0,0 +1,204 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:web_socket/web_socket.dart';
+
+import 'cupertino_api.dart';
+
+/// An error occurred while connecting to the peer.
+class ConnectionException extends WebSocketException {
+  final Error error;
+
+  ConnectionException(super.message, this.error);
+
+  @override
+  String toString() => 'CupertinoErrorWebSocketException: $message $error';
+}
+
+/// A [WebSocket] implemented using the
+/// [NSURLSessionWebSocketTask API](https://developer.apple.com/documentation/foundation/nsurlsessionwebsockettask).
+class CupertinoWebSocket implements WebSocket {
+  /// Create a new WebSocket connection using the
+  /// [NSURLSessionWebSocketTask API](https://developer.apple.com/documentation/foundation/nsurlsessionwebsockettask).
+  ///
+  /// The URL supplied in [url] must use the scheme ws or wss.
+  ///
+  /// If provided, the [protocols] argument indicates that subprotocols that
+  /// the peer is able to select. See
+  /// [RFC-6455 1.9](https://datatracker.ietf.org/doc/html/rfc6455#section-1.9).
+  static Future<CupertinoWebSocket> connect(Uri url,
+      {Iterable<String>? protocols, URLSessionConfiguration? config}) async {
+    if (!url.isScheme('ws') && !url.isScheme('wss')) {
+      throw ArgumentError.value(
+          url, 'url', 'only ws: and wss: schemes are supported');
+    }
+
+    final readyCompleter = Completer<CupertinoWebSocket>();
+    late CupertinoWebSocket webSocket;
+
+    final session = URLSession.sessionWithConfiguration(
+        config ?? URLSessionConfiguration.defaultSessionConfiguration(),
+        // In a successful flow, the callbacks will be made in this order:
+        // onWebSocketTaskOpened(...)        // Good connect.
+        // <receive/send messages to the peer>
+        // onWebSocketTaskClosed(...)        // Optional: peer sent Close frame.
+        // onComplete(..., error=null)       // Disconnected.
+        //
+        // In a failure to connect to the peer, the flow will be:
+        // onComplete(session, task, error=error):
+        //
+        // `onComplete` can also be called at any point if the peer is
+        // disconnected without Close frames being exchanged.
+        onWebSocketTaskOpened: (session, task, protocol) {
+      webSocket = CupertinoWebSocket._(task, protocol ?? '');
+      readyCompleter.complete(webSocket);
+    }, onWebSocketTaskClosed: (session, task, closeCode, reason) {
+      assert(readyCompleter.isCompleted);
+      webSocket._connectionClosed(closeCode, reason);
+    }, onComplete: (session, task, error) {
+      if (!readyCompleter.isCompleted) {
+        // `onWebSocketTaskOpened should have been called and completed
+        // `readyCompleter`. So either there was a error creating the connection
+        // or a logic error.
+        if (error == null) {
```

Comment:
This could be an assert. You can pass a string message as the second arg:
`assert(error != null, 'expected an error or "onWebSocketTaskOpened" to be called first');`

Reply:
I left it this way because `assert()` isn't a `null` guard i.e. I'd have to `error!` on line 74.

Let me know if you'd prefer that.

Reply:
Oh, right. That makes sense.